lib/Platform/Intl/PlatformIntlICU.cpp

```diff
@@ -120,40 +419,782 @@ double Collator::compare(
+        skeleton += u"HH";
+    }
+  }
+
+  if (minute_.has_value()) {
+    if (minute_ == eNumeric)
+      skeleton += u"m";
+    else if (minute_ == eTwoDigit)
+      skeleton += u"mm";
+  }
+
+  if (second_.has_value()) {
+    if (second_ == eNumeric)
+      skeleton += u"s";
+    else if (second_ == eTwoDigit)
+      skeleton += u"ss";
+  }
+
+  UErrorCode status = U_ZERO_ERROR;
+  std::u16string bestpattern;
+  int32_t patternLength;
+
+  UDateTimePatternGenerator *dtpGenerator = udatpg_open(&locale8_[0], &status);
+  patternLength = udatpg_getBestPatternWithOptions(
+      dtpGenerator,
+      &skeleton[0],
+      -1,
+      UDATPG_MATCH_ALL_FIELDS_LENGTH,
+      nullptr,
+      0,
+      &status);
+
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    bestpattern.resize(patternLength);
+    udatpg_getBestPatternWithOptions(
+        dtpGenerator,
+        &skeleton[0],
+        skeleton.length(),
+        UDATPG_MATCH_ALL_FIELDS_LENGTH,
+        &bestpattern[0],
+        patternLength,
+        &status);
+  }
+
+  // if timezone is specified, use that instead, else use default
+  if (!timeZone_.empty()) {
+    const UChar *timeZoneRes =
+        reinterpret_cast<const UChar *>(timeZone_.c_str());
+    int32_t timeZoneLength = timeZone_.length();
+    return udat_open(
+        UDAT_PATTERN,
+        UDAT_PATTERN,
+        &locale8_[0],
+        timeZoneRes,
+        timeZoneLength,
+        &bestpattern[0],
+        patternLength,
+        &status);
+  } else {
+    return udat_open(
+        UDAT_PATTERN,
+        UDAT_PATTERN,
+        &locale8_[0],
+        nullptr,
+        -1,
+        &bestpattern[0],
+        patternLength,
+        &status);
+  }
+}
+
+vm::CallResult<std::u16string> DateTimeFormatICU::getDefaultHourCycle(
+    vm::Runtime &runtime) {
+  UErrorCode status = U_ZERO_ERROR;
+  std::u16string myString;
+  // open the default UDateFormat and Pattern of locale
+  UDateFormat *defaultDTF = udat_open(
+      UDAT_DEFAULT,
+      UDAT_DEFAULT,
+      &locale8_[0],
+      nullptr,
+      -1,
+      nullptr,
+      -1,
+      &status);
+  int32_t size = udat_toPattern(defaultDTF, true, nullptr, 0, &status);
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    myString.resize(size + 1);
+    udat_toPattern(defaultDTF, true, &myString[0], 40, &status);
+    assert(status <= 0); // Check for errors
+    udat_close(defaultDTF);
+    // find the default hour cycle and return it
+    for (int32_t i = 0; i < size; i++) {
+      char16_t ch = myString[i];
+      switch (ch) {
+        case 'K':
+          return u"h11";
+          break;
+        case 'h':
+          return u"h12";
+          break;
+        case 'H':
+          return u"h23";
+          break;
+        case 'k':
+          return u"h24";
+          break;
+      }
+    }
+  }
+
+  // There should always be a default hour cycle, return an exception if not
+  return vm::ExecutionStatus::EXCEPTION;
```

Comment:
This is incorrect, and may result in a crash if we hit this case. If we need to raise an exception here, we should do it with `runtime.raiseError` as we do in other places in this file.